src/distributions/binomial.rs

```diff
@@ -77,53 +65,80 @@ impl Distribution<u64> for Binomial {
             1.0 - self.p
         };
 
-        // prepare some cached values
-        let float_n = self.n as f64;
-        let ln_fact_n = log_gamma(float_n + 1.0);
-        let pc = 1.0 - p;
-        let log_p = p.ln();
-        let log_pc = pc.ln();
-        let expected = self.n as f64 * p;
-        let sq = (expected * (2.0 * pc)).sqrt();
-
-        let mut lresult;
-
-        // we use the Cauchy distribution as the comparison distribution
-        // f(x) ~ 1/(1+x^2)
-        let cauchy = Cauchy::new(0.0, 1.0);
-        loop {
-            let mut comp_dev: f64;
+        let result;
+
+        // For small n * min(p, 1 - p), the BINV algorithm based on the inverse
+        // transformation of the binomial distribution is more efficient:
+        //
+        // Voratas Kachitvichyanukul and Bruce W. Schmeiser. 1988. Binomial
+        // random variate generation. Commun. ACM 31, 2 (February 1988),
+        // 216-222. http://dx.doi.org/10.1145/42372.42381
+        if (self.n as f64) * p < 10. {
+            let q = 1. - p;
+            let s = p / q;
+            let a = ((self.n + 1) as f64) * s;
+            let mut r = q.powf(self.n as f64);
+            // FIXME: Using integer exponentiation should be faster, but it is not available for
+            // `u64` exponents, neither in `std` or `num`.
```

Comment:
That would be a nice fix for the integer exponentiation issue. However, it requires some more thought to determine an appropriate cutoff. Maybe looking at other implementations helps with that.

Reply:
It seems like GSL does not have an `n` cutoff for the BINV algorithm. However, they are using a cutoff for the BINV loop. If it is exceeded, [they sample a new random value and start over](https://github.com/ampl/gsl/blob/644e768630841bd085cb7121085a688c4ff424d0/randist/binomial_tpe.c#L124).

Reply:
I'm not sure whether a cutoff for large `n` is a good idea. After all, the issue triggering this PR (#734) was about performance problems for large `n` and very small `p`.

Reply:
@geq1t would you like to comment?

I don't know (without analysis), but the paper mentioned use of 64-bit types in one impl to avoid rounding errors, implying the internals might need higher precision than the input type.

The reason I suggested the cut-off is because it sounds like it avoids this problem, though of course it might be far more restrictive than necessary. Perhaps we should simply use this as-is for now but flag it as needing review for accuracy under certain conditions.